std-bits/table/src/main/java/org/enso/table/data/table/join/SortJoin.java

```diff
@@ -0,0 +1,136 @@
+  private final Storage<?>[] lowerStorages;
+  private final Storage<?>[] upperStorages;
+
+  @Override
+  public JoinResult join(ProblemAggregator problemAggregator) {
+    Context context = Context.getCurrent();
+
+    int leftRowCount = conditionsHelper.getLeftTableRowCount();
+    List<OrderedMultiValueKey> leftKeys = new ArrayList<>(leftRowCount);
+    for (int i = 0; i < leftRowCount; i++) {
+      leftKeys.add(new OrderedMultiValueKey(leftStorages, i, directions));
+      context.safepoint();
+    }
+
+    NavigableSet<OrderedMultiValueKey> leftIndex = buildSortedLeftIndex(leftKeys);
+    JoinResult.Builder resultBuilder = new JoinResult.Builder();
+
+    if (leftIndex.isEmpty()) {
+      return resultBuilder.build();
+    }
+
+    int rightRowCount = conditionsHelper.getRightTableRowCount();
+    for (int rightRowIx = 0; rightRowIx < rightRowCount; rightRowIx++) {
+      addMatchingLeftRows(leftIndex, rightRowIx, resultBuilder);
+      context.safepoint();
+    }
+
+    return resultBuilder.build();
+  }
+
+  @Override
+  public void joinSubsets(
+      List<Integer> leftGroup,
+      List<Integer> rightGroup,
+      JoinResult.Builder resultBuilder,
+      ProblemAggregator problemAggregator) {
+    Context context = Context.getCurrent();
+
+    List<OrderedMultiValueKey> leftKeys =
+        leftGroup.stream().map(i -> new OrderedMultiValueKey(leftStorages, i, directions)).toList();
+    NavigableSet<OrderedMultiValueKey> leftIndex = buildSortedLeftIndex(leftKeys);
+
+    if (leftIndex.isEmpty()) {
+      return;
+    }
+
+    for (int rightRowIx : rightGroup) {
+      addMatchingLeftRows(leftIndex, rightRowIx, resultBuilder);
+      context.safepoint();
+    }
+  }
+
+  private NavigableSet<OrderedMultiValueKey> buildSortedLeftIndex(List<OrderedMultiValueKey> keys) {
+    Context context = Context.getCurrent();
+    TreeSet<OrderedMultiValueKey> index = new TreeSet<>();
+    for (var key : keys) {
+      context.safepoint();
+      if (key.hasAnyNulls()) {
+        continue;
+      }
+      index.add(key);
+    }
+    return index;
+  }
+
+  private OrderedMultiValueKey buildLowerBound(int rightRowIx) {
+    return new OrderedMultiValueKey(lowerStorages, rightRowIx, directions);
+  }
+
+  private OrderedMultiValueKey buildUpperBound(int rightRowIx) {
+    return new OrderedMultiValueKey(upperStorages, rightRowIx, directions);
+  }
+
+  void addMatchingLeftRows(
+      NavigableSet<OrderedMultiValueKey> sortedKeys,
+      int rightRowIx,
+      JoinResult.Builder resultBuilder) {
+    OrderedMultiValueKey lowerBound = buildLowerBound(rightRowIx);
+    OrderedMultiValueKey upperBound = buildUpperBound(rightRowIx);
+
+    // If the match interval is invalid or empty, there is nothing to do.
+    if (lowerBound.hasAnyNulls()
+        || upperBound.hasAnyNulls()
+        || lowerBound.compareTo(upperBound) > 0) {
+      return;
+    }
+
+    NavigableSet<OrderedMultiValueKey> firstCoordinateMatches =
```

Comment:
How does this only match on the first coordinate? Doesn't it use `OrderedMultiValueKey.compareTo`, which checks all fields? Just trying to understand the approach.

Reply:
Hmm, maybe I should describe it more clearly, I understand where the confusion is coming from.

So yes, we are comparing by all coordinates - this is a bit of a heuristic that may make the algorithm _slightly_ faster in some cases.

The thing is, when we sort lexicographically what we get is chunks grouped by the first coordinate, and _then_ within each chunk we have it sorted along second coordinate etc.

Let's do an example. Let's take a list of pairs sorted lexicographically:
```
[-100, 0]; [0, 1]; [1, 1]; [1, 4]; [1, 300]; [2, 1]; [2, 20]; [3, 10]; [10, 20]
```

If we search for `lowerBound = [1, 1]` and `upperBound = [2, 2]`, then when we do the `subSet` on this list we will get a sub-list of entries between these two in this lexicographic order:
```
[1, 1]; [1, 4]; [1, 300]; [2, 1]
```

But you can see that e.g. `[1, 300]` is not really a valid match - it's there because it is between 2 valid first coordinate matches: `[1, 1]` and `[2, 1]`. So that's why I call these only first coordinate matches and I actually have to filter them out to ensure only the valid matches are kept.